src/common/convenience.h

```diff
@@ -0,0 +1,161 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include <mutex>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include <boost/thread/shared_mutex.hpp>
+
+#include "common/backport14.h"
+#include "common/shunique_lock.h"
+
+#include "include/assert.h" // I despise you. Not you the reader, I'm talking
+                            // to the include file.
+
+
+#ifndef CEPH_COMMON_CONVENIENCE_H
```

Comment:
no and yes. i understand that `#pragma` offers implementation defined behavior, so does `#pragma once`, and it [is not likely](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0538r0.html#why-not-reuse-pragma) to be a part of the standard. but since it's widely supported, so i guess we can use it. but i agree, that we should stick to `#ifdef` if we need to be more standard compliant.